### PR TITLE
fix(float): position fix when relative is win

### LINKED ISSUE #14735: nvim_open_win with relative=win doesn't work if win is a float
When you use `nvim_open_win` with `relative=win` and point `win=` to a floating window, then the new float will be relative to the original non-floating window.

### Steps to reproduce using `nvim -u NORC`

```lua
local function popup(win, bg)
	win = win or vim.api.nvim_get_current_win()
	local buf = vim.api.nvim_create_buf(false, true)
	local ret = vim.api.nvim_open_win(buf, true, {
		relative = "win",
		win = win,
		col = 4,
		row = 4,
		width = 20,
		height = 10,
	})
	vim.api.nvim_win_set_option(ret, "winhl", "Normal:" .. bg .. ",EndOfBuffer:" .. bg)
	return ret
end

vim.cmd([[hi Popup1 guibg=red]])
vim.cmd([[hi Popup2 guibg=yellow]])

local p1 = popup(nil, "Popup1")
local p2 = popup(p1, "Popup2")
```

### Actual behaviour

Both the yellow and red floating window will be shown at the same place

### Expected behaviour

The second yellow window should be positioned relative to the first one, with an offset of 4 rows and 4 columns

### Other weird stuff

* When doing a vsplit of the current window, and then running the code when the second window is active, you'll see that:
  - the red popup opens in the right window
  -  the yellow popup opens in the left window (still at the wrong position)

* Probably related to #13403: `nvim_win_get_position` always returns `[0,0]` for floats
* `vim.api.nvim_win_get_number` of the floating window, returns the window number of the window where the float was created.

### PR DESCRIPTION
**Pull request by glepnir**
Problem: current win ui render order will cause float window postion in wrong position.

Solution: render from last.

Fix #14735

### CODE REVIEW

File: src/nvim/window.c [hunk 1]
```diff
@@ -7614,18 +7614,20 @@ void win_get_tabwin(handle_T id, int *tabnr, int *winnr)
 
 void win_ui_flush(bool validate)
 {
-  FOR_ALL_TAB_WINDOWS(tp, wp) {
-    if (wp->w_pos_changed && wp->w_grid_alloc.chars != NULL) {
+  FOR_ALL_TABS(tp) {
+    for (win_T *wp = ((tp) == curtab) ? lastwin : (tp)->tp_lastwin; wp != NULL; wp = wp->w_prev){
```
Comment (justinmk) [inline src/nvim/window.c:7618]:
```suggestion
    // Iterate in _reverse_ order.
    for (win_T *wp = ((tp) == curtab) ? lastwin : (tp)->tp_lastwin; wp != NULL; wp = wp->w_prev){
```

File: test/functional/ui/float_spec.lua [hunk 2]
```diff
@@ -291,8 +287,41 @@ describe('float window', function()
       return vim.api.nvim_win_get_position(win_id)
     ]])
 
-    eq(14, pos[1])
-    eq(12, pos[2])
+    eq({14,12}, pos)
+  end)
+
+  it('win render order issue #14735', function()
```
Comment (justinmk) [inline test/functional/ui/float_spec.lua:293]:
```suggestion
  it('relative=win is relative to given (floating) window #14735', function()
```

**Review by justinmk — APPROVED** [review]
LGTM after comments, assuming CI passes.

File: src/nvim/window.c [hunk 3]
```diff
@@ -961,7 +961,7 @@ void ui_ext_win_position(win_T *wp, bool validate)
   FloatConfig c = wp->w_float_config;
   if (!c.external) {
     ScreenGrid *grid = &default_grid;
-    Float row = c.row, col = c.col;
+    int row = (int)c.row, col = (int)c.col;
```
Comment (zeertzjq) [inline src/nvim/window.c:964]: Revert

**Review by zeertzjq — CHANGES_REQUESTED** [review]

**Comment by justinmk** [comment]
uncrustify is flagging something in src/nvim/window.c

**Review by zeertzjq — CHANGES_REQUESTED** [review]
I don't think this is the correct solution. If an older floating window is reconfigured to be relative to a newer floating window then this won't work.

**Comment by glepnir** [comment]
change relative to new window ? hmm let me try reproduce it ...

**Comment by zeertzjq** [comment]
This works fine without this PR, but is broken with this PR
```lua
local function test_nvim()
  local parent_bufnr = vim.api.nvim_create_buf(false, true)
  local bufnr = vim.api.nvim_create_buf(false, true)

  local initial_opts = {
    width = 5,
    height = 5,
    row = 5,
    col = 5,
    relative = "editor",
    style = "minimal",
  }
  local win_id = vim.api.nvim_open_win(bufnr, false, initial_opts)
  local parent_winid = vim.api.nvim_open_win(parent_bufnr, false, initial_opts)

  local parent_window_config = {
    width = 10,
    height = 6,
    row = 10,
    col = 20,
    relative = "editor",
    style = "minimal",
    border = "rounded",
    zindex = 49,
  }

  vim.api.nvim_win_set_config(parent_winid, parent_window_config)

  vim.api.nvim_win_set_option(parent_winid, "winhighlight", "Normal:Normal")

  local opts = {
    width = 8,
    height = 4,
    row = 1,
    col = 1,
    relative = "win",
    style = "minimal",
    win = parent_winid,
    zindex = 50,
  }

  vim.api.nvim_win_set_config(win_id, opts)

  return vim.api.nvim_win_get_position(win_id)
end

test_nvim()
```

**Comment by glepnir** [comment]
thanks.

**Comment by glepnir** [comment]
hmm seems to be doing the same thing. Add grid recursively

**Pull request closed without merging**